src/accountsdb/buffer_pool.zig

```diff
@@ -0,0 +1,1074 @@
+            if (alive_eviction_attempts < 10 or alive_eviction_attempts % 2 == 0) {
+                if (maybe_evicted == null) maybe_evicted = self.evictMain(metadata);
+                if (maybe_evicted == null) maybe_evicted = self.evictSmall(metadata);
+            } else {
+                if (maybe_evicted == null) maybe_evicted = self.evictSmall(metadata);
+                if (maybe_evicted == null) maybe_evicted = self.evictMain(metadata);
+            }
+
+            const evicted = maybe_evicted orelse
+                @panic("unable to evict: cache empty");
+
+            // alive evicted keys are reinserted, we try again
+            if (metadata.rc[evicted].isAlive()) {
+                metadata.freq[evicted] = 1;
+                self.main.writeItemAssumeCapacity(evicted);
+                metadata.in_queue[evicted] = .main;
+                alive_eviction_attempts += 1;
+                continue;
+            }
+
+            // key is definitely dead
+            metadata.in_queue[evicted] = .none;
+            break evicted;
+        };
+
+        return dead_key;
+    }
+
+    fn evictGhost(self: *HierarchicalFIFO, metadata: Metadata) ?Key {
+        const evicted: ?Key = while (self.ghost.readItem()) |ghost_key| {
+            switch (metadata.in_queue[ghost_key]) {
+                .ghost => {
+                    break ghost_key;
+                },
+                .main => {
+                    // This key has moved from ghost to main, we will just pop
+                    // and ignore it.
+                },
+                .none => {
+                    // This key moved from ghost to main, and then was evicted
+                    // from main. However, because ghost is always evicted from
+                    // before main, this should not be possible.
+                    unreachable;
+                },
+                .small => unreachable,
+            }
+        } else null;
+
+        return evicted;
+    }
+
+    fn evictMain(self: *HierarchicalFIFO, metadata: Metadata) ?Key {
+        const evicted: ?Key = while (self.main.readItem()) |main_key| {
+            if (metadata.in_queue[main_key] != .main) unreachable;
+
+            if (metadata.freq[main_key] == 0) {
+                break main_key;
+            } else {
+                metadata.freq[main_key] -|= 1;
+                self.main.writeItemAssumeCapacity(main_key);
+            }
+        } else null;
+
+        return evicted;
+    }
+
+    fn evictSmall(self: *HierarchicalFIFO, metadata: Metadata) ?Key {
+        const evicted: ?Key = while (self.small.readItem()) |small_key| {
+            if (metadata.in_queue[small_key] != .small) unreachable;
+
+            if (metadata.freq[small_key] == 0) {
+                break small_key;
+            } else {
+                metadata.freq[small_key] -|= 1;
+                self.small.writeItemAssumeCapacity(small_key);
+            }
+        } else null;
+
+        return evicted;
+    }
```

Comment:
small and main are both the same type and these methods do the same - should just be one method?

Reply:
merged https://github.com/Syndica/sig/pull/368/commits/1c8f4bdcfe4295dac4f14acabd0ae3af413c169e#diff-9334997573b7124ee92d8ed1efaa5a70b34f8531988349743900d1cf39014569L643-R700